ckanext/datagovuk/tests/test_s3_resources_plugin.py

```diff
@@ -0,0 +1,150 @@
+        mock_toolkit.get_action = MockAction
+
+        plugin = DatagovukPlugin()
+        resource = self.resource.copy()
+
+        plugin.before_create_or_update({}, resource)
+        assert not mock_abort.called
+        assert config.get("ckan.datagovuk.s3_url_prefix") in resource["url"]
+
+    @patch("boto3.resource")
+    @patch("ckanext.datagovuk.upload.toolkit.abort")
+    @patch("ckanext.datagovuk.upload.toolkit")
+    def test_before_create_or_update_empty_upload(
+        self, mock_toolkit, mock_abort, mock_resource
+    ):
+        mock_toolkit.get_action = MockAction
+
+        plugin = DatagovukPlugin()
+        resource = self.resource.copy()
+        resource["upload"] = ""
+
+        plugin.before_create_or_update({}, resource)
+        assert not mock_abort.called
+        assert config.get("ckan.datagovuk.s3_url_prefix") not in resource["url"]
+
+    @patch("boto3.resource")
+    @patch("ckanext.datagovuk.upload.toolkit.abort")
+    @patch("ckanext.datagovuk.upload.toolkit")
+    def test_before_create_or_update_undefined_upload(
+        self, mock_toolkit, mock_abort, mock_resource
+    ):
+        mock_toolkit.get_action = MockAction
+
+        plugin = DatagovukPlugin()
+        resource = self.resource.copy()
+        del resource["upload"]
+
+        plugin.before_create_or_update({}, resource)
+        assert not mock_abort.called
+        assert config.get("ckan.datagovuk.s3_url_prefix") not in resource["url"]
+
+    @patch("boto3.resource")
+    @patch("ckanext.datagovuk.upload.toolkit.abort")
+    @patch("ckanext.datagovuk.upload.toolkit")
+    def test_before_create_or_update_upload_api(
+        self, mock_toolkit, mock_abort, mock_resource
+    ):
+        mock_toolkit.get_action = MockAction
+
+        plugin = DatagovukPlugin()
+        resource = self.resource.copy()
+        resource["format"] = "API"
+
+        plugin.before_create_or_update({}, resource)
+        assert not mock_abort.called
+        assert config.get("ckan.datagovuk.s3_url_prefix") not in resource["url"]
+
+    @patch("boto3.resource")
+    @patch("ckanext.datagovuk.upload.toolkit.abort")
+    @patch("ckanext.datagovuk.upload.toolkit")
+    def test_upload_early_abort(self, mock_toolkit, mock_abort, mock_resource):
+        mock_toolkit.get_action = MockAction
+
+        plugin = DatagovukPlugin()
+        resource = self.resource.copy()
+        resource["upload"] = "dummy data"  # *not* a FieldStorage
+
+        plugin.before_create_or_update({}, resource)
+        assert not mock_abort.called
+        assert config.get("ckan.datagovuk.s3_url_prefix") not in resource["url"]
+
+    @patch("ckanext.datagovuk.upload.toolkit.abort")
+    @patch("ckanext.datagovuk.upload.toolkit")
+    def test_upload_s3_exception(self, mock_toolkit, mock_abort):
+        mock_toolkit.get_action = MockAction
+
+        s3 = botocore.session.get_session().create_client("s3")
```

Comment:
Why not just mock out `put_object` and introduce a `side_effect` so that an exception is raised? This should make the test a bit clearer.